### ISSUE TITLE
basic questions about ActiveJ Serializer. overflow, public field, annotation, platform independent, thread safe.

### ISSUE THREAD

**Issue by lifeinwild** [activej/activej#147]
ActiveJ Serializer is extreme fast in my JMH benchmark:
```
# JMH version: 1.34
# VM version: JDK 1.8.0_312, OpenJDK 64-Bit Server VM, 25.312-b07

Benchmark                                   Mode  Cnt      Score   Error  Units
BenchmarkActiveJ.deserialize               thrpt    2  20456.246          ops/s
BenchmarkActiveJ.serialize                 thrpt    2  43678.163          ops/s
BenchmarkKryo.deserialize                  thrpt    2  10062.984          ops/s
BenchmarkKryo.serialize                    thrpt    2  13294.549          ops/s
BenchmarkKryoCustomSerializer.deserialize  thrpt    2  15841.880          ops/s
BenchmarkKryoCustomSerializer.serialize    thrpt    2  36915.721          ops/s
```
no code because of copyright of JMH by Oracle.

I have some questions.

1. kryo supports mix buffer of byte[] and ByteArrayOutputStream. So you can design without the risk of overflow with good performance of fixed buffer. How about ActiveJ Serializer?
     ```
        private static final ByteArrayOutputStream bufferKryo = new ByteArrayOutputStream();
        private static final Output outputKryo = new Output(bufferKryo,
			SerializerBenchmarkCommon.bufferSize);
     ```

     ```
		if (bufferKryo.size() != 0) {
			outputKryo.flush();
			serializedKryo = bufferKryo.toByteArray();
		} else {
			serializedKryo = outputKryo.toBytes();
		}
      ```
2. ActiveJ Serializer requests public field. Is there a way to handle private fields?
3. ActiveJ Serializer requests annotations and full constructor. There is a risk of forgetting to write annotations and full constructor. any practice?
4. If ActiveJ Serializer serialized equivalent objects in different environment for each, does it create equivalent byte array in any environment?
5. Is ActiveJ Serializer pure bytecode?
6. Is BinarySerializer thread safe?

**Comment by eduard-vasinskyi** [comment]
Hi, @lifeinwild

1. In order to serialize the data without the risk of overflow ActiveJ Serializer has a [StreamCodec](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/stream/StreamCodec.java) interface (and its implementation [StreamCodec#ofBinarySerializer](https://github.com/activej/activej/blob/665ca09399cf0938cd970ce11963039651c7ff8c/core-serializer/src/main/java/io/activej/serializer/stream/StreamCodec.java#L23-L29)). A `StreamCodec` allows writing serialized data directly to an `OutputStream` (or read from an `InputStream`). Take a look at [StreamInputOutputTest](https://github.com/activej/activej/blob/master/core-serializer/src/test/java/io/activej/serializer/StreamInputOutputTest.java) for usage reference.
2. Yes you can serialize private fields, you just have to add public getters with [@Serialize](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/annotations/Serialize.java) annotations. (as in this [example](https://activej.io/serializer/examples#generics-and-interfaces))
3. Proper annotations are required for a correct ActiveJ Serializer performance. If you miss any annotation or misconfigure a serialized class, you will get a runtime exception.
4. Yes, the byte arrays of the serialized data are equal. Also, note that if you want to guarantee that serialized data is the same across different JVMs, make sure to specify a unique field `order` parameter on `@Serialize` annotation. You can skip it if you only plan to use a serializer within the same JVM.
5. Yes, ActiveJ Serializer does not use `Unsafe`
6. Every implementation of a [BinarySerializer](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/BinarySerializer.java) interface in ActiveJ is thread safe.

**Comment by lifeinwild** [comment]
Thank you for your reply.
I have further questions.

1. After serializing in StreamCodec, deserializing in byte[] way failed.
    Is the binary format different depending on the serialization method?
    ```
    @Test
	public void streamTest() throws IOException {
		BinarySerializer<String> serializer = SerializerBuilder.create()
				.build(String.class);
		StreamCodec<String> codec = StreamCodec.ofBinarySerializer(serializer);

		String before = "abc";

		ByteArrayOutputStream baos = new ByteArrayOutputStream();
		try (StreamOutput streamOutput = StreamOutput.create(baos, 1)) {
			codec.encode(streamOutput, before);
		}

		byte[] serialized = baos.toByteArray();
		ByteArrayInputStream bais = new ByteArrayInputStream(serialized);
		try (StreamInput streamInput = StreamInput.create(bais)) {
			String deserialized1 = codec.decode(streamInput);
			assertEquals(before, deserialized1);
			String deserialized2 = serializer.decode(serialized, 0);
			assertEquals(before, deserialized2);//failed
		}
	}
    ```
2. Is it mandatory to add @Serialize to interface getter?
    https://activej.io/serializer/examples#generics-and-interfaces
    ```
    public interface Person<K, V> {
        @Serialize
        List<Skill<K, V>> getSkills();
    }
    ```

3. Is there a way to collectively mark up all getters for serialization for a class and its child classes?
4. What is the meaning of SerializeClass and SerializeClasses? Is there a list of all annotation descriptions?
5. Could ActiveJ Serializer deserialize a byte[] that doesn't know which class it came from? 
    in other words, does the serialized byte[] contain information about the class?
6. Is there an annotation for size limit to a variable-length field such as String or List?
7. HashMap does not guarantee order. 
    LinkedHashMap is suitable for persistent objects because it guarantees order.
    
    ActiveJ Serializer supports HashMap, but not LinkedHashMap.
    https://github.com/activej/activej/tree/master/core-serializer/src/main/java/io/activej/serializer/impl
    Do I need to write a custom serializer for LinkedHashMap?

**Comment by eduard-vasinskyi** [comment]
1. `StreamCodec` as the name suggests is actually meant for serialization of a stream of data items. To be able to deserialize individual data items from a stream, it needs to understand where serialized data item ends (in other words, to implement message framing, which is a ‘self-delimiting’ data encoding). `StreamCodec` does this by prepending 1-4 byte header which is a varint that represents the size of the following message bytes itself array. This is similar to how protobuf works (see the explanation on [StackOverflow](https://stackoverflow.com/questions/52336629/what-are-the-delimiters-for-protobuf-messages/52348530#52348530)).
If you do not want a self-delimited serialized data representation you can opt out of using a `StreamCodec` and implement it by passing a large enough array directly to a `BinarySerializer` (or increase its size iteratively [as its actually done in a StreamCodec](https://github.com/activej/activej/blob/665ca09399cf0938cd970ce11963039651c7ff8c/core-serializer/src/main/java/io/activej/serializer/stream/StreamCodecs.java#L536-L543))
2. In this example it is not necessary to add `@Serialize` annotation to an interface getter. This is because we pass the `Developer.class` to the `SerializerBuilder#build` method, so `SerializerBuilder` will scan the Developer class for annotations. However, if we wanted to create a `BinarySerializer` for an interface Person and passed `Person.class` to the `SerializerBuilder#build` method, we would have to add the `@Serialize` annotation to an interface getter.
3. You can add `@Serialize` annotations to a parent class or an interface and then build a `BinarySerializer` for that parent class/interface.
4. A `@SerializeClass` annotation allows you to pass a [SerializerDef](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/SerializerDef.java) class to be used to serialize a specific class field or a type. It also allows you to specify which classes this field/type can be an instance of. For example you can mark a field of type `Object` to be a subclass of either a `String` or an `Integer` class. 
A `@SerializeClasses` is a helper annotation that makes `@SerializeClasses` annotation repeatable. See [documentation](https://docs.oracle.com/javase/tutorial/java/annotations/repeating.html).
We will add an explanation on serializer annotation to our website. For now you can read the existing documentation/examples [here](https://activej.io/serializer). You can also explore ActiveJ Serializer tests, especially [BinarySerializerTest](https://github.com/activej/activej/blob/master/core-serializer/src/test/java/io/activej/serializer/BinarySerializerTest.java) as a usage reference.
5. A serialized byte array does not contain information about the serialized class. However, you can use subclass serialization to define which classes can a certain field/type be an instance of. This way a serialized byte array would additionally contain an index which represents the class that is serialized. Take a look at this [example](https://github.com/activej/activej/blob/master/examples/core/serializer/src/main/java/SerializeSubclassesExample.java) that demonstrates a serialization of a List of arbitrary types. You can also look at [BinarySerializerTest](https://github.com/activej/activej/blob/master/core-serializer/src/test/java/io/activej/serializer/BinarySerializerTest.java) for usages of `@SerializeClass` annotation with specified `subclasses` or `subclassesId` fields.
6. ActiveJ Serializer does fully support `LinkedHashMap` serialization. You can find all of the default serializers for Java collections [here](https://github.com/activej/activej/blob/665ca09399cf0938cd970ce11963039651c7ff8c/core-serializer/src/main/java/io/activej/serializer/SerializerBuilder.java#L159-L174). All you need to do is specify a type of a `@Serialize` annotated field/getter as a `LinkedHashMap`.
You can also take a look at [this test ](https://github.com/activej/activej/blob/665ca09399cf0938cd970ce11963039651c7ff8c/core-serializer/src/test/java/io/activej/serializer/BinarySerializerTest.java#L2442)for map serialization as a reference

**Comment by lifeinwild** [comment]
I apologize my misundersting about LinkedHashMap.

1. https://github.com/activej/activej/blob/master/examples/core/serializer/src/main/java/SerializeSubclassesExample.java
    Does serializer.decode() line 44 depend on subClasses line 33?
    In other words, for example, in the communication process, 
    if sender sends a serialized byte[] and receiver deserializes it,
    Do they need to share the information like subClasses?

2. Is it necessary to write grandchild class name on @SerializeClass? 
    Is the whitelist of subclasses for security?

  ```
public class AbstractClassTest {
	@Test
	public void test() {
		BinarySerializer<SampleAbstractClass> serializer = SerializerBuilder
				.create().build(SampleAbstractClass.class);
		byte[] buffer = new byte[1000];
		SampleAbstractClass before = new SampleClass("abc");
		int size = serializer.encode(buffer, 0, before);
		byte[] serialized = new byte[size];
		System.arraycopy(buffer, 0, serialized, 0, size);
		SampleAbstractClass deserialized = serializer.decode(serialized, 0);
		assertEquals(before.getS(), deserialized.getS());
		assertEquals(before.getClass(), deserialized.getClass());
	}

	/**
	 * grandchild class name SampleClass.
	 */
	@SerializeClass(
			subclasses = { SampleAbstractClass2.class, SampleClass.class })
	public static abstract class SampleAbstractClass {
	}

	@SerializeClass(subclasses = SampleClass.class)
	public static abstract class SampleAbstractClass2
			extends SampleAbstractClass {
		abstract public String getS();
	}

	public static class SampleClass extends SampleAbstractClass2 {
		public String s;

		public SampleClass(@Deserialize("s") String s) {
			this.s = s;
		}

		@Override
		@Serialize
		public String getS() {
			return s;
		}
	}
}
  ```

**Comment by eduard-vasinskyi** [comment]
1. Yes. Basically, the serializer writes a first byte as a subclass index, then subclass data. The deserializer reads the index, and dispatches to corresponding subclass deserializer.
That’s why a deserializing `BinarySerializer` must be configured exactly the same way as the serializing `BinarySerializer`, and order of classes in `SerializerBuilder#withSubclasses` must match.
2. Subclasses must be concrete instantiatable classes. It does not make sense to specify an interface or an abstract class (`SampleAbstractClass2.class` in your case) as a subclass. This is because a proper subclass is chosen based on class equality.
So, in your example you do not need to specify `SampleAbstractClass2.class` as a subclass. And as the result you do need to put `@SerializeClass` annotation on `SampleAbstractClass2 class` (unless you plan to build a serializer for the `SampleAbstractClass2`, rather than `SampleAbstractClass`)
 
_PS. If you want to dive deeper into ActiveJ Serializer you can enable `DefiningClassLoader#withDebugOutputDir` option and then pass the resulting `DefiningClassLoader` to the `SerializerBuilder#create` method. This will write all of the generated `BinarySerializer` classes to the specified directory. It is useful for debugging and for better understanding of how ActiveJ Serializer works.
PSS. I have noticed that you copy the serialized byte array before deserialization. This is not necessary, you can pass the original array (`buffer`) to the `BinarySerializer#decode` method. A `BinarySerializer` will read only the required amount of bytes from the byte array._

**Comment by lifeinwild** [comment]
I read that class file Object_1. 

1. Are the class indexes unique only in @SerializeClass?
    Can class indexes be unique in all BinarySerializers that is created by same SerializerBuilder?
    In other words, can I set the mapping between class and index to SerializerBuilder rather than model classes?
        
    Thus each class is mapped to the same index wherever it appears across classes and fields, like kryo's Registration and ClassResolver. 
    
    In my limited understanding, such an option makes it easy to use because we don't need @SerializeClass and withSubclasses() in such mapping.

**Comment by lifeinwild** [comment]
About the StringToInt lookup, I compared the performance of the huge switch statement, HashMap, and LinkedHashMap in JMH. The test is a function that converts String to int by 1000 mappings.

```
Benchmark                           Mode  Cnt         Score   Error  Units
JDKHashMap.testStringIntMap        thrpt    2  38430978.141          ops/s
JDKLinkedHashMap.testStringIntMap  thrpt    2  36663783.374          ops/s
Switch.test                        thrpt    2   5895096.748          ops/s
```

HashMap is seven times faster than switch statement. 

Guessing from some benchmark results I've written before, **despite the performance of HashMap, it is not enough to outperform ActiveJ Serializer**.

In other words, **ActiveJ Serializer achieves better performance than HashMap by the idea of little distributed lookuptables**. So it use class index instead of class ID.

I understand that ActiveJ Serializer design decisions are one of the best. Thank you for replying many times.
